src/curp-command.proto

```diff
@@ -37,18 +30,75 @@ message CmdResult {
+// This error will be serialized in the `detail` of the gRPC Status and returned
+// to clients. The reason for not dividing into some levels is for the
+// convenience of the client to make judgments. If some errors never be checked
+// in client, they could be merged into `Internal`.
+//
+// Feel free to use `google.protobuf.Empty` as the client never concerned the
+// details happens in the server. If your error needs the context to be passed
+// within the server, you can implement a separate error and implement the
+// From<YourError> trait for `CurpError`, see `StorageError`. Or your error
+// context may also be passed with the client, add your error details in
+// `CurpError`, see `Redirect` below.
+//
+// Please add comments above your error if you want to extend more :)
+message CurpError {
+    message Redirect {
+        optional uint64 leader_id = 1;
+        uint64 term = 2;
+    }
+    oneof err {
+        // When a command conflicts with the server SP.
+        google.protobuf.Empty KeyConflict = 1;
+        // When a proposal id is duplicated.
+        google.protobuf.Empty Duplicated = 2;
+        // When the client id in a proposal id is expired.
+        google.protobuf.Empty ExpiredClientId = 3;
+        // When the conf change config is invalid.
+        google.protobuf.Empty InvalidConfig = 4;
+        // When the conf change proposal remove a node which is not exists.
+        google.protobuf.Empty NodeNotExists = 5;
+        // When the conf change proposal add a node which is exists.
+        google.protobuf.Empty NodeAlreadyExists = 6;
+        // When the conf change proposal promote a lazy learner.
+        google.protobuf.Empty LearnerNotCatchUp = 7;
+        // When the cluster is shutting down
+        google.protobuf.Empty ShuttingDown = 8;
+        // When the request cluster_version does not match the cluster's.
+        google.protobuf.Empty WrongClusterVersion = 9;
+        // When rpc transport error happens, this means the request does not send
+        // to server
+        google.protobuf.Empty RpcTransport = 10;
+
+        // When the request should be send to leader and this node is not the
+        // leader. Return the context to client to update its state.
+        Redirect redirect = 11;
+
+        // The internal error that the client never concerned.
+        // e.g.
+        // 1. failed to allocate a new snapshot
+        // 2. failed to reset the command executor by snapshot
+        // 3. failed to get last applied index from command executor.
+        // 4. io::Error
+        // 5. failed to encode and decode some stuff, like pb serialize or bincode
+        // serialize.
+        // 6. failed in storage layer
+        // ...
+        string Internal = 12;
+    }
+}
+
+message ProposeRequest {
+    // The serialized command
+    // The original type is `Command`
+    bytes command = 1;
+    uint64 cluster_version = 2;
 }
 
 message ProposeResponse {
```

Comment:
Why should we remove the `leader_id` and `term` in the `ProposeResponse`?

Reply:
I'll try to add later in `ProposeRequest`